File: src/main/java/util/notepad/MacNotepad.java
package util.notepad;

import lombok.extern.log4j.Log4j2;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Paths;
import java.util.Objects;

@Log4j2
public class MacNotepad implements Notepad {

    public void openNotepad() {
      String scriptPath = "";
      try {
        scriptPath = Paths.get(Objects.requireNonNull(MacNotepad.class.getClassLoader().getResource("openNotepad.scpt")).toURI()).toString();
        ProcessBuilder pb = new ProcessBuilder("osascript", scriptPath);
        pb.start().waitFor();
      } catch (IOException | InterruptedException e) {
        e.printStackTrace();
        Thread.currentThread().interrupt();
      } catch (URISyntaxException e) {
        //noinspection StringConcatenationArgumentToLogCall bc of throwable
        log.error("Error getting the script path (" + scriptPath + ")", e);
      }
    }

    public void writeText(String text) {
        // TODO: Implement this method
    }

    public void deleteText() {
        // TODO: Implement this method
    }

    public void addNewLine() {
        // TODO: Implement this method
    }

    public void saveFileAs(String name) {
        // TODO: Implement this method
    }

    public void openNewFile() {
        // TODO: Implement this method
    }

    public void closeFile() {
        // TODO: Implement this method
    }

    public void closeNotepad() {
        // TODO: Implement this method
    }
}

File: src/main/kotlin/util/notepad/MacNotepad.kt
package util.notepad

import util.extension.command
import util.extension.control
import util.extension.enter
import util.extension.type
import java.awt.Robot
import java.awt.event.KeyEvent
import java.io.IOException

class MacNotepad : Notepad {
    private val robot = Robot()
    private var process: Process? = null

    override fun openNotepad() {
        try {
            process = ProcessBuilder("open", "-a", "Notes").start()
            Thread.sleep(2000) // allow app to open before typing
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun deleteText() {
        robot.control(KeyEvent.VK_A)
        robot.type(KeyEvent.VK_BACK_SPACE)
    }

    override fun addNewLine() {
        robot.enter()
    }

    override fun saveFileAs(name: String) {
        robot.command(KeyEvent.VK_S)
    }

    override fun openNewFile() {
        try {
            ProcessBuilder(
                "osascript",
                "-e",
                """
                tell application \"System Events\"
                    click menu item \"New Note\" of menu 1 of menu bar item \"File\" of menu bar 1 of application \"Notes\"
                end tell
                """.trimIndent(),
            ).start()
                .waitFor()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun closeNotepad() {
        try {
            ProcessBuilder("osascript", "-e", "tell application \"Notes\" to quit").start().waitFor()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
